strinject.c

```diff
@@ -0,0 +1,462 @@
+	{STRINJECT_TYPE_FILE_LINE, "INJECT_FILE_LINE"},
+	{STRINJECT_TYPE_UNKNOWN, NULL}
+};
+
+/*
+ * string inject functions
+ */
+
+static struct {
+	enum RTR_STRINJECT_FUNC_ID id;
+	const char *name;
+} g_inject_funcs[] = {
+	{STRINJECT_FUNC_FWRITE, "fwrite"},
+	{STRINJECT_FUNC_FREAD, "fread"},
+	{STRINJECT_FUNC_READ, "read"},
+	{STRINJECT_FUNC_WRITE, "write"},
+	{STRINJECT_FUNC_SEND, "send"},
+	{STRINJECT_FUNC_SENDTO, "sendto"},
+	{STRINJECT_FUNC_SENDMSG, "sendmsg"},
+	{STRINJECT_FUNC_RECV, "recv"},
+	{STRINJECT_FUNC_RECVFROM, "recvfrom"},
+	{STRINJECT_FUNC_RECVMSG, "recvmsg"},
+	{STRINJECT_FUNC_MAX, NULL}
+};
+
+/*
+ * string injection init function
+ */
+
+static void rtr_strinject_init(void)
+{
+	RTR_CONFIG_HANDLE config = RTR_CONFIG_START;
+
+	while (1) {
+		char *inject_type_str = NULL;
+		char *func_list = NULL;
+		char *inject_param = NULL;
+		double inject_rate;
+
+		enum RTR_STRINJECT_TYPE inject_type = STRINJECT_TYPE_UNKNOWN;
+
+		int i, reverse;
+
+		/* get configuration line */
+		if (rtr_get_config_multiple(&config, "stringinject", ARGUMENT_TYPE_STRING, ARGUMENT_TYPE_STRING,
+			ARGUMENT_TYPE_STRING, ARGUMENT_TYPE_DOUBLE, ARGUMENT_TYPE_END,
+			&inject_type_str, &func_list, &inject_param, &inject_rate) == 0)
+			break;
+
+		/* get inject type */
+		for (i = 0; i < STRINJECT_TYPE_UNKNOWN; i++) {
+			if (real_strcmp(inject_type_str, g_inject_types[i].type_str) == 0) {
+				inject_type = g_inject_types[i].type;
+				break;
+			}
+		}
+
+		if (inject_type == STRINJECT_TYPE_UNKNOWN)
+			continue;
+
+		/* get function list */
+		for (i = 0; i < STRINJECT_FUNC_MAX; i++) {
+			if (rtr_check_config_token(g_inject_funcs[i].name, func_list, "|", &reverse) &&
+				!g_strinject_infos[i].exist) {
+				g_strinject_infos[i].type = inject_type;
+				real_strcpy(g_strinject_infos[i].param, inject_param);
+				g_strinject_infos[i].rate = inject_rate;
+
+				g_strinject_infos[i].exist = 1;
+			}
+		}
+	}
+
+	/* set init flag */
+	g_strinject_init = 1;
+}
+
+/*
+ * parse injection parameter
+ */
+
+static int parse_inject_param(enum RTR_STRINJECT_TYPE type, const char *param, size_t len, void *val, off_t *offset)
+{
+	char param_val[512];
+	char *p;
+
+	int fmt_count, overflow_buf_len;
+
+	/* parse param */
+	p = real_strchr(param, ':');
+	if (!p)
+		return -1;
+
+	memset(param_val, 0, sizeof(param_val));
+	real_strncpy(param_val, param, p - param);
+
+	if (real_strcmp(p + 1, "RANDOM") == 0)
+		*offset = rand() % len;
+	else {
+		*offset = strtol(p + 1, NULL, 10);
+		if (*offset == LONG_MIN || *offset == LONG_MAX)
+			return -1;
+	}
+
+	/* check offset */
+	if (*offset >= len)
+		*offset = len - 1;
+
+	/* inject string by types */
+	switch (type) {
+	case STRINJECT_TYPE_HEX:
+		if (real_strcmp(param_val, "RANDOM") == 0)
+			((char *) val)[0] = rand() % 0xFF;
+		else if (real_strncmp(param_val, "0x", 2) == 0) {
+			char hex_val;
+
+			hex_val = strtol(param_val + 2, NULL, 16);
+			if (hex_val == LONG_MIN || hex_val == LONG_MAX)
+				return -1;
```

Comment:
Test is a little pointless as hex_val is a char so is unlikely to be LONG_{MIN,MAX}. I'd use
```
unsigned long hex_val;
hex_val = strtoul(...);
if (hex_val > 0xff)
    return -1;
((char *)val)[0] = (char)hex_val;
```